### PR TITLE
feat: vm.toString()

### PR DESCRIPTION
**Pull request by odyslam**
## Motivation

Facilitate conversion to strings to be used in other cheat codes (e.g `--ffi`, `--filewrite`).

Some conversions (e.g bytes -> string) are not trivial in native solidity.

## Solution

- Add the cheat code
- Use UIfmt that already exists by moving it to foundry_common

### CODE REVIEW

**Comment by onbjerg** [comment]
Looks ok to me - wonder why we use `UiFmt` as opposed to the `Display` traits in ethers as a general thought? Also you have some merge conflicts to resolve :smile:

**Comment by odyslam** [comment]
For the use of `UIfmt` I would defer to @mattsse

**Comment by mattsse** [comment]
> wonder why we use UiFmt as opposed to the Display traits in ethers as a general thought?

mainly because all `Display` impls in eth types truncate... so this way we have control over what's printed

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 1]
```diff
@@ -2,14 +2,15 @@ use super::Cheatcodes;
...
+    abi::{AbiEncode, Address, Token},
+    prelude::{k256::ecdsa::SigningKey, Lazy, LocalWallet, Signer, H160},
     types::{NameOrAddress, H256, U256},
     utils,
     utils::keccak256,
 };
-use once_cell::sync::Lazy;
+use foundry_common::fmt::*;
 use revm::{CreateInputs, Database, EVMData};
+use std::str;
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/util.rs:13]: this should not be needed

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 2]
```diff
@@ -71,6 +72,24 @@ pub fn apply<DB: Database>(
...
+        }
+        HEVMCalls::ToString4(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString5(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString6(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/util.rs:92]: this is indeed weird, need to debug this in `abigen!`

**Review by mattsse — CHANGES_REQUESTED** [review]
so far so good, looks like imports need to be updated for test to pass

File: evm/src/executor/inspector/cheatcodes/util.rs (hunk 2 above)
Comment (odyslam) [reply evm/src/executor/inspector/cheatcodes/util.rs:92]: using expand, it indeed produces 0 to 6, skipping 1.

**Comment by odyslam** [comment]
@mattsse let's merge this?

**Review by mattsse — CHANGES_REQUESTED** [review]
lgtm,

needs nightly formatting `cargo +nightly fmt` and clippy fixes

also the missing `HEVMCalls::ToString1` has been fixed on ethers so you can
`cargo update -p ethers` and update the match statement accordingly

**Comment by odyslam** [comment]
@mattsse addressed all your comments.  I tried building locally, but I still encountered the issue. 

How can I make sure that I am building with the latest version of ethers? Also, what was the issue? I tried to follow the PR a bit, but I am not certain I fully understand the bug. 

Thanks!

**Comment by onbjerg** [comment]
@odyslam Run (in root):

```
cargo update -p ethers
```

After this the bug where `HEVMCalls::ToString1` is skipped is fixed so you need to update your cheatcode stuff

```
cargo +nightly fmt --all
```

And then you'd need clippy fixes.

**Comment by onbjerg** [comment]
Pushed the changes so we can get it merged :smile:

File: anvil/core/src/eth/transaction/mod.rs [hunk 3]
```diff
@@ -962,7 +962,7 @@ mod tests {
             TypedTransaction::Legacy(tx) => tx,
             _ => panic!("Invalid typed transaction"),
         };
-        assert_eq!(tx.input, b"".into());
+        assert_eq!(tx.input, Into::<Bytes>::into(b""));
```
Comment (mattsse) [inline anvil/core/src/eth/transaction/mod.rs:965]:
```suggestion
        assert_eq!(tx.input, Bytes::from(b""));
```

**Review by mattsse — APPROVED** [review]

**Pull request merged**